torch/functional.py

```diff
@@ -989,16 +1009,35 @@ def tensordot(a, b, dims=2, out=None):
     """
     if has_torch_function_variadic(a, b):
         return handle_torch_function(tensordot, (a, b), a, b, dims=dims)
-    if isinstance(dims, (list, tuple)) or \
-       (isinstance(dims, torch.Tensor) and dims.numel() > 1):
+
+    dims_a: List[int] = []
+    dims_b: List[int] = []
+
+    if isinstance(dims, (tuple, list)):
         dims_a, dims_b = dims
-    else:
-        if isinstance(dims, torch.Tensor):
-            dims = dims.item()
+
+    if isinstance(dims, torch.Tensor):
+        num_elements = dims.numel()
+        if num_elements > 1:
+            assert dims.size()[0] == 2
+            dims_a = torch.jit.annotate(List[int], dims[0].tolist())
+            dims_b = torch.jit.annotate(List[int], dims[1].tolist())
+        else:
+            dims_val = int(dims.item())
+            if dims_val < 0:
+                raise RuntimeError(f"tensordot expects dims >= 0, but got dims={dims}")
+            dims_a = list(range(-dims_val, 0))
+            dims_b = list(range(dims_val))
+
+    if isinstance(dims, int):
         if dims < 0:
             raise RuntimeError(f"tensordot expects dims >= 0, but got dims={dims}")
         dims_a = list(range(-dims, 0))
         dims_b = list(range(dims))
+
+    if len(dims_a) == 0 or len(dims_b) == 0:
+        raise RuntimeError(f"unsupported input to tensordot, got dims={dims}")
```

Comment:
This check [breaks](https://github.com/pyro-ppl/pyro/issues/2886) the simple case `tensordot(torch.tensor(0.), torch.tensor(0.), 0)`, which works fine in NumPy
```py
>>> np.tensordot(np.zeros(()), np.zeros(()), 0)
array(0.)
```
This important edge case `dims=0` is needed by generic machinery as in [Pyro](https://github.com/pyro-ppl/pyro/blob/5ecc9c0a2b8192b99ad386cc1d946e8a44f964ae/pyro/infer/util.py#L303) and [opt_einsum](https://github.com/dgasmith/opt_einsum/blob/be2d3dcb9792016d3b7ed0afacfed70024698c9e/opt_einsum/backends/torch.py#L54).  Can we revert this check?

cc @neerajprad